lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs

```diff
@@ -117,12 +123,35 @@ changeBalance = foldl' addCoin 0 . change
 feeBalance :: CoinSelection -> Word64
 feeBalance sel = inputBalance sel - outputBalance sel - changeBalance sel
 
+-- | Total UTxO balance + withdrawal.
+totalBalance :: Quantity "lovelace" Word64 -> [(TxIn, TxOut)] -> Word64
+totalBalance (Quantity withdraw) inps = balance' inps + withdraw
+
 addTxOut :: Integral a => a -> TxOut -> a
 addTxOut total = addCoin total . coin
 
 addCoin :: Integral a => a -> Coin -> a
 addCoin total c = total + (fromIntegral (getCoin c))
 
+-- | Compute the fraction of the first input to match the given ratio, rounded
+-- down.
+--
+-- /invariant:/ The ratio mustn't be greater than 1 without risk to overflow.
+--
+-- >>> 10 `proportionallyTo` 1%1
+-- 10
+--
+-- >>> 10 `proportionallyTo` 1%2
+-- 5
+--
+-- >>> 10 `proportionallyTo` 1%3
+-- 3
+proportionallyTo :: Integral a => a -> Ratio a -> a
```

Comment:
any chance to have unit test for this or something that uses this inside?

Reply:
ok, so it is used with `totalWithdraw` and checked in random...

Reply:
it is checked below

Reply:
Indeed!